### PR TITLE
fix(consul): unable to recreate after canceled the consul watch

### PR DESCRIPTION
**Pull request by harbourlga**

### CODE REVIEW

**Comment by harbourlga** [comment]
agree

File: contrib/registry/consul/registry.go [hunk 1]
```diff
@@ -209,12 +209,11 @@ func (r *Registry) Watch(ctx context.Context, name string) (registry.Watcher, er
...
-	if !ok {
-		err := r.resolve(ctx, set)
-		if err != nil {
-			return nil, err
-		}
+	err := r.resolve(ctx, set)
+	if err != nil {
+		return nil, err
 	}
+
```
Comment (shenqidebaozi) [inline contrib/registry/consul/registry.go:216]:
```go
if err := r.resolve(ctx, set); err != nil {
     return nil, err
}
```

Comment (harbourlga) [reply contrib/registry/consul/registry.go]: 已修改

**Comment by shenqidebaozi** [comment]
Can it be fixed when multiple clients are created for the same service? For example, if a client is idle, will b client also be unable to update the instance due to Watcher shutdown?

**Comment by harbourlga** [comment]
> Can it be fixed when multiple clients are created for the same service? For example, if a client is idle, will b client also be unable to update the instance due to Watcher shutdown?
> 
> https://github.com/go-kratos/kratos/blob/0cb74ea2a330b1a91ca892a158a041d8925becad/contrib/registry/consul/watcher.go#L34

If you are talking about multiple grpc clients, then each client will build a different resolver to watch name resolution updates. Each resolver has its own context to control the shutdown of the Watcher

https://github.com/grpc/grpc-go/blob/02858ee5064040458f7e3e04e10765a62814c50b/resolver_wrapper.go#L80

https://github.com/go-kratos/kratos/blob/0cb74ea2a330b1a91ca892a158a041d8925becad/transport/grpc/resolver/discovery/builder.go#L85

https://github.com/go-kratos/kratos/blob/0cb74ea2a330b1a91ca892a158a041d8925becad/transport/grpc/resolver/discovery/builder.go#L88

**Comment by harbourlga** [comment]
> Can it be fixed when multiple clients are created for the same service? For example, if a client is idle, will b client also be unable to update the instance due to Watcher shutdown?
> 
> https://github.com/go-kratos/kratos/blob/0cb74ea2a330b1a91ca892a158a041d8925becad/contrib/registry/consul/watcher.go#L34

If you are talking about multiple grpc clients, then each client will build a different resolver to watch name resolution updates. Each resolver has its own context to control the shutdown of the Watcher. Therefore, when client a becomes idle, client b will not close the watcher.

When creating the grpc client, a resolver will be built: 
https://github.com/grpc/grpc-go/blob/02858ee5064040458f7e3e04e10765a62814c50b/resolver_wrapper.go#L80
https://github.com/go-kratos/kratos/blob/0cb74ea2a330b1a91ca892a158a041d8925becad/transport/grpc/resolver/discovery/builder.go#L86

https://github.com/go-kratos/kratos/blob/0cb74ea2a330b1a91ca892a158a041d8925becad/transport/grpc/resolver/discovery/builder.go#L88

**Review by shenqidebaozi — APPROVED** [review]

**Review by daemon365 — APPROVED** [review]

**Pull request merged**